### ISSUE TITLE
java.lang.IllegalStateException: Reply already submitted with simultaneous method calls

### ISSUE THREAD

**Issue by crazecoder** [flutter/flutter#29092]
## Steps to Reproduce

 flutter
```
Future<Null> avoid() async {
    await _channel.invokeMethod('a', map);
}
Future<Null> bvoid() async {
    await _channel.invokeMethod('b', map);
}
```
android
```
@Override
    public void onMethodCall(final MethodCall call, final Result result) {
        if (call.method.equals("a")) {
             result.success("result a");
        } else if(call.method.equals("b")) {
             result.success("result b");
        } else {
             result.notImplemented();
        }
    }
```
example
```
avoid();
bvoid();
```

## Logs

```
 java.lang.IllegalStateException: Reply already submitted
        at io.flutter.view.FlutterNativeView$PlatformMessageHandlerImpl$1.reply(FlutterNativeView.java:197)
        at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:204)
```
An error is reported when two methods are executed at the same time

**Comment by dnfield** [comment]
I suspect this isn't just an issue on Android, but tagging it Android since it's reported for Android.

I'm not sure if we can easily support this use case, but maybe @mklim or @bkonyi have better ideas.

**Comment by devxpy** [comment]
@dnfield May I suggest something? 

Can we just catch this, suppress it, and just print a message in the logs, instead of crashing the whole app?

In my experience, most often than not, it's a stupid callback that gets called multiple times, which in turn results in `result.success` being called multiple times, and crashing the whole app, when in reality the desired result was already sent back to flutter the first time around :/

**Comment by dnfield** [comment]
I suspect we actually need some kind of identifier to go with the request and response here, or perhaps a counter.  Swallowing this seems like it could cause difficult to diagnose problems.

**Comment by neiljaywarner** [comment]
Is it odd that I am getting this in release mode and I wasn't getting it in debug mode?  does that mean it's like a race condition...

**Comment by dnfield** [comment]
It is likely a race condition, but it could also mean there's a bug that's AOT vs JIT specific.

**Comment by hatemragab** [comment]
i got this error when use real time pusher java.lang.IllegalStateException: Reply already submitted

**Comment by neiljaywarner** [comment]
> It is likely a race condition, but it could also mean there's a bug that's AOT vs JIT specific.

Any thoughts on how to determine which it is if i have time? I'm going to open another ticket for it because it is not quite the same...

**Comment by SAGARSURI** [comment]
I am trying to build a messaging app. I am using method channels to send and receive messages. First time when I receive a message its working perfectly i.e passing the message from android to Flutter view using the MethodChannel. But the second it crashes resulting:
 
`java.lang.IllegalStateException: Reply already submitted`

Here is my android code:

`call.method == "receiveMessage" -> receiveMessage(result)`

```
private fun receiveMessage(result: MethodChannel.Result){
        val listenerID="com.sagar.gossip.MainActivity"

        MyChat.addMessageListener(listenerID, object : MyChat.MessageListener() {
            override fun onTextMessageReceived(p0: TextMessage?) {
                Log.d(TAG, "Message received successfully: " + p0?.toString())
                result.success(p0?.text)
            }
        })
    }
```

My Flutter code:
```
Future<String> receiveMessage() async {
    String message = '';
    try{
      message = await platform.invokeMethod('receiveMessage');
    } on PlatformException catch(e){
      message = e.message;
    }
    return message;
  }
```

**Comment by devxpy** [comment]
As a temporary fix, I've released a [flutter plugin](https://github.com/pycampers/method_call_dispatcher) that tries to ingest this error.

The exception is muted [here](https://github.com/pycampers/method_call_dispatcher/blob/38da44e4e6016a2da684d67b10df7eef51dd3a47/android/src/main/kotlin/com/pycampers/method_call_dispatcher/MethodCallDispatcherPlugin.kt#L36).

**Comment by elmarou** [comment]
I want to use the Authentication Registration feature with a Google Account, but when the function (_singInWithGoogle) is called, the application stops and displays the following error.
***ERROR***
```D/AndroidRuntime(11262): Shutting down VM 
E/AndroidRuntime(11262): FATAL EXCEPTION: main 
E/AndroidRuntime(11262): Process: com.hayriyad.erpstage, PID: 11262 
E/AndroidRuntime(11262): java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=53293, result=0, data=null} to activity {com.hayriyad.erpstage/com.hayriyad.erpstage.MainActivity}: java.lang.IllegalStateException: Reply already submitted E/AndroidRuntime(11262): at android.app.ActivityThread.deliverResults(ActivityThread.java:4360)
E/AndroidRuntime(11262): at android.app.ActivityThread.handleSendResult(ActivityThread.java:4402) E/AndroidRuntime(11262): at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:49) 
```
***i use this fonction***
```void _signInWithGoogle() async {
    final GoogleSignInAccount googleUser = await _googleSignIn.signIn();
    final GoogleSignInAuthentication googleAuth =
    await googleUser.authentication;
    final AuthCredential credential = GoogleAuthProvider.getCredential(
      accessToken: googleAuth.accessToken,
      idToken: googleAuth.idToken,

    );
    final FirebaseUser user = await _auth.signInWithCredential(credential);
    assert(user.email != null );
    assert(user.displayName != null);
    assert(!user.isAnonymous);
    assert(await user.getIdToken() != null);

    final FirebaseUser currentUser = await _auth.currentUser();
    assert(user.uid == currentUser.uid);
    setState(() {
      if (user != null) {
        _success = true;
        _userID = user.uid;
      } else {
        _success = false;
      }
    });
  }```

**Comment by devxpy** [comment]
Does a similar issue exist on ios?

**Comment by Karlheinzniebuhr** [comment]
same issue

`D/AndroidRuntime(30555): Shutting down VM
E/AndroidRuntime(30555): FATAL EXCEPTION: main
E/AndroidRuntime(30555): Process: com.example.TODOer, PID: 30555
E/AndroidRuntime(30555): java.lang.IllegalStateException: Reply already submitted
E/AndroidRuntime(30555): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:124)
E/AndroidRuntime(30555): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:204)
E/AndroidRuntime(30555): 	at com.lyokone.location.LocationPlugin$2.onLocationResult(LocationPlugin.java:281)
E/AndroidRuntime(30555): 	at com.google.android.gms.internal.location.zzau.notifyListener(Unknown Source:4)
E/AndroidRuntime(30555): 	at com.google.android.gms.common.api.internal.ListenerHolder.notifyListenerInternal(Unknown Source:17)
E/AndroidRuntime(30555): 	at com.google.android.gms.common.api.internal.ListenerHolder$zaa.handleMessage(Unknown Source:5)
E/AndroidRuntime(30555): 	at android.os.Handler.dispatchMessage(Handler.java:105)
E/AndroidRuntime(30555): 	at android.os.Looper.loop(Looper.java:164)
E/AndroidRuntime(30555): 	at android.app.ActivityThread.main(ActivityThread.java:6944)
E/AndroidRuntime(30555): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(30555): 	at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
E/AndroidRuntime(30555): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
I/zygote64(30555): Background concurrent copying GC freed 11907(1065KB) AllocSpace objects, 0(0B) LOS objects, 65% free, 815KB/2MB, paused 11.645ms total 46.643ms`

**Comment by jaceshim** [comment]
same issue!
E/AndroidRuntime(19943): FATAL EXCEPTION: main
E/AndroidRuntime(19943): Process: nomad9.unsego.user, PID: 19943
E/AndroidRuntime(19943): java.lang.IllegalStateException: Reply already submitted
E/AndroidRuntime(19943): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:126)
E/AndroidRuntime(19943): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:225)
E/AndroidRuntime(19943): 	at nomad9.unsego.user.MainActivity$3.onSuccess(MainActivity.java:199)
E/AndroidRuntime(19943): 	at nomad9.unsego.user.MainActivity$3.onSuccess(MainActivity.java:142)
E/AndroidRuntime(19943): 	at com.kakao.network.callback.ResponseCallback.onSuccessForUiThread(ResponseCallback.java:76)
E/AndroidRuntime(19943): 	at com.kakao.network.tasks.KakaoResultTask$1$1.run(KakaoResultTask.java:78)
E/AndroidRuntime(19943): 	at android.os.Handler.handleCallback(Handler.java:873)
E/AndroidRuntime(19943): 	at android.os.Handler.dispatchMessage(Handler.java:99)
E/AndroidRuntime(19943): 	at android.os.Looper.loop(Looper.java:214)
E/AndroidRuntime(19943): 	at android.app.ActivityThread.main(ActivityThread.java:7045)
E/AndroidRuntime(19943): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(19943): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/AndroidRuntime(19943): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:965)
I/Process (19943): Sending signal. PID: 19943 SIG: 9
---------------------------------------------------------------
❯ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel unknown, v1.7.4, on Mac OS X 10.14.5 18F132, locale en-KR)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 10.2.1)
[✓] iOS tools - develop for iOS devices
[✓] Android Studio (version 3.3)
[✓] IntelliJ IDEA Ultimate Edition (version 2018.3.2)
[✓] VS Code (version 1.35.1)
[✓] Connected device (1 available)

• No issues found!

------------------

**Comment by bradyt** [comment]
Same.

```
E/MethodChannel#cone.tangential.info/uri( 4515): Failed to handle method call
E/MethodChannel#cone.tangential.info/uri( 4515): java.lang.IllegalStateException: Reply already submitted
E/MethodChannel#cone.tangential.info/uri( 4515): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:124)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.error(MethodChannel.java:209)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at info.tangential.cone.MainActivity.isUriOpenable(MainActivity.java:144)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at info.tangential.cone.MainActivity.access$000(MainActivity.java:35)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at info.tangential.cone.MainActivity$1.onMethodCall(MainActivity.java:57)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:201)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:88)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:219)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at android.os.MessageQueue.next(MessageQueue.java:323)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at android.os.Looper.loop(Looper.java:136)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at android.app.ActivityThread.main(ActivityThread.java:6186)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
E/MethodChannel#cone.tangential.info/uri( 4515): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
```

**Comment by bradyt** [comment]
I believe my issue was resolved by making sure I use `try catch finally` properly to close my Java `InputStream`.

**Comment by moonljt521** [comment]
我也遇到了这个问题，是因为在android端用MethodChannel.Result 回调的时候，在switch case 语句里，丢失了break语句，大概是多次Result多次调用引起的崩溃，我补上了break，这个问题便被解决了

**Comment by moonljt521** [comment]
I also encountered this problem when callbacked on the android side using the Method Channel Result.It is because that the “break statement” is lost in the switch case statement.It probably that the multiple “Results”and multiple “Calls” caused the crash. I made up the “break”, then the problem was solved.

**Comment by jigarpatel17** [comment]
@moonljt521 Can you please share more details, or probably, code snippet will also helpful. Thanks

**Comment by bradyt** [comment]
> I believe my issue was resolved by making sure I use `try catch finally` properly to close my Java `InputStream`.

Oh, I made a similar mistake again during a refactor. Glad I found this. Really need to make sure my code does not try to return a result twice.

**Comment by FangWW** [comment]
Avoid multiple calls to a "result".

**Comment by PolashPaul2367** [comment]
same issue

E/AndroidRuntime(19977): Caused by: java.lang.IllegalStateException: Reply already submitted
E/AndroidRuntime(19977): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:151)
E/AndroidRuntime(19977): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:225)
E/AndroidRuntime(19977): 	at vn.hunghd.flutter.plugins.imagecropper.ImageCropperDelegate.onActivityResult(ImageCropperDelegate.java:106)
E/AndroidRuntime(19977): 	at io.flutter.app.FlutterPluginRegistry.onActivityResult(FlutterPluginRegistry.java:204)
E/AndroidRuntime(19977): 	at io.flutter.app.FlutterActivityDelegate.onActivityResult(FlutterActivityDelegate.java:132)
E/AndroidRuntime(19977): 	at io.flutter.app.FlutterActivity.onActivityResult(FlutterActivity.java:142)
E/AndroidRuntime(19977): 	at android.app.Activity.dispatchActivityResult(Activity.java:8924)
E/AndroidRuntime(19977): 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4422)
E/AndroidRuntime(19977): 	... 11 more

**Comment by longravuth** [comment]
For anyone who is facing this issue because you are calling result.success() or result.error() from a native callback function, i found the solution. Implement your request handler like in CameraRequestPermissionsListener class in https://github.com/flutter/plugins/blob/master/packages/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java 

then when you receive the callback in your methodhandler class, simple set result = null after you call result.success or result.failure to reset its state like processResult() method in
https://github.com/BaseflowIT/flutter-permission-handler/blob/develop/android/src/main/java/com/baseflow/permissionhandler/PermissionHandlerPlugin.java


:<

**Comment by 103style** [comment]
**if you call result.success() or result.xxx()  twice,  you will get this problem！**

```
public class MainActivity extends FlutterActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        GeneratedPluginRegistrant.registerWith(this);

        new MethodChannel(getFlutterView(), CHANNEL)
                .setMethodCallHandler(
                        (methodCall, result) -> {
                            if (methodCall.method.equals(METHOD_NAME)) {
                                int batteryLevel = 0;
                                result.success(batteryLevel);
                               // result.success(batteryLevel);   
                               // if you call result.success() or result.xxx()  twice,  you will get this problem！   

                            } else {
                                result.notImplemented();
                            }
                        });
    }
}
```

**Comment by hatemragab** [comment]
Yes

**Comment by Stone2517** [comment]
Hey, guys, I solved Reply already submitted problem by adding a break to my switch statement like a return for if-else statement.

**Comment by DhavalRKansara** [comment]
**Yeah guys, I found the solution for this check if you call result.success() or result.xxx() twice, you will get this problem！**

**Comment by tabarakMohammed** [comment]
if (registrar.activity() == null) {
      // If a background flutter view tries to register the plugin, there will be no activity from the registrar,
      // we stop the registering process immediately because requires an activity.
      return;
    }

**Comment by OmkarSsawant** [comment]
> I am trying to build a messaging app. I am using method channels to send and receive messages. First time when I receive a message its working perfectly i.e passing the message from android to Flutter view using the MethodChannel. But the second it crashes resulting:
> 
> `java.lang.IllegalStateException: Reply already submitted`
> 
> Here is my android code:
> 
> `call.method == "receiveMessage" -> receiveMessage(result)`
> 
> ```
> private fun receiveMessage(result: MethodChannel.Result){
>         val listenerID="com.sagar.gossip.MainActivity"
> 
>         MyChat.addMessageListener(listenerID, object : MyChat.MessageListener() {
>             override fun onTextMessageReceived(p0: TextMessage?) {
>                 Log.d(TAG, "Message received successfully: " + p0?.toString())
>                 result.success(p0?.text)
>             }
>         })
>     }
> ```
> 
> My Flutter code:
> 
> ```
> Future<String> receiveMessage() async {
>     String message = '';
>     try{
>       message = await platform.invokeMethod('receiveMessage');
>     } on PlatformException catch(e){
>       message = e.message;
>     }
>     return message;
>   }
> ```

Getting same error

**Comment by efreibe** [comment]
@OmkarSsawant. Because you are calling addMessageListener every time you execute receiveMessage, you ended calling result.success multiple times. That is why you have this error. 
Try to add the listener one time only.

**Comment by OmkarSsawant** [comment]
> @OmkarSsawant. Because you are calling addMessageListener every time you execute receiveMessage, you ended calling result.success multiple times. That is why you have this error.
> Try to add the listener one time only

> @OmkarSsawant. Because you are calling addMessageListener every time you execute receiveMessage, you ended calling result.success multiple times. That is why you have this error.
> Try to add the listener one time only.

But In my case I am replying the message in onActivityResultListner in which it occurs only once .when i try to submit result my crashes with error result already submitted .
When actually i have not submitted it

**Comment by Kelier** [comment]
So what should I do if I want to call the plugin multiple times and return different values?

**Comment by efreibe** [comment]
@OmkarSsawant Yes. But you are adding a listener every time you call the plugin, so, the first time it fires ok, the second call it fires twice, the third call it fires three times and so on. Try removing the listener before sending the result back.

@Kelier MethodChannel allows you to make one call and receive one result at a time. You can call it any number of times and wil receive a result every time it was called.
If you want to receive multiple results, like events, then you will need to implement an EventChannel. See https://api.flutter.dev/flutter/services/EventChannel-class.html

**Comment by OmkarSsawant** [comment]
The `_result` is just assisgned in onmethodcall method
```
@Override
    public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_CODE) {

            AuthorizationResponse response = AuthorizationClient.getResponse(resultCode, data);
            if (response.getType() == TOKEN) {
                // Handle successful response
                final String accessToken = response.getAccessToken();
              _result.success( accessToken);
             
                return false;
            } else if (response.getType() == ERROR) {
               _result.error("",responce.getError(),null);
                
                return false;
            }
        }
        return false;
    }

```



I get same error

**Comment by efreibe** [comment]
Could be related to this issue #48202?

**Comment by anakin-jin** [comment]
it's wired.
following code (in android side) will cause the same problem:

  ```
@Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    }
    if (call.method.equals("getDHKeys")) {
      result.success( getDHPGK() );
    }
    else {
      result.notImplemented();
    }
  }
```

after adding a word 'else' before second 'if' sentence. the problem fixed.

 ```
 @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    }
    else if (call.method.equals("getDHKeys")) {//!!!! here! added a word 'else'!!!!
      result.success( getDHPGK() );
    }
    else {
      result.notImplemented();
    }
  }
```

the flutter side code:

```
      platformVersion = await Libblufi.platformVersion;
      print("ver:$platformVersion");

```

**Comment by xster** [comment]
Seems like a pretty annoying problem. @gaaclarke can you take a look to see if there are concurrency problems with our platform channel implementation?

**Comment by gaaclarke** [comment]
@crazecoder @xster This works for me.  Here is an [example app that does it](https://github.com/gaaclarke/double_result/).  Here is the [double calls](https://github.com/gaaclarke/double_result/blob/397c643eabac458db81fe1b3ba01a9337aef3472/example/lib/main.dart#L40).  I tested it on:
```
[✓] Flutter (Channel unknown, 1.18.0-14.0.pre.3, on Mac OS X 10.15.4 19E287, locale en-US)
    • Flutter version 1.18.0-14.0.pre.3 at /Users/aaclarke/dev/flutter
    • Framework revision e6b697a9df (5 days ago), 2020-05-07 11:03:05 -0700
    • Engine revision d6aa099de7
    • Dart version 2.9.0 (build 2.9.0-5.0.dev 4da5b40fb6)
```

I suspect there is extra steps required to reproduce this issue that aren't enumerated in the issue description.  Feel free to use my code as an example of what you want to do.  Let me know if you think there is something I missed.

**Comment by IkhwanSI13** [comment]
So this error happened when you use native code like Java right? In my case, I didn't type a single code like result.success or another result. I just type native code for Firebase Cloud Messaging. Maybe this error comes from the packages that I used?
So how to solve this? Btw, I get this error from Firebase Crashlytics

**Comment by gaaclarke** [comment]
@IkhwanSI13 if you are getting the error and you aren't implementing method channels I'd audit the plugin's code to see if they are doing something like https://github.com/flutter/flutter/issues/29092#issuecomment-619784521

**Comment by yeahia2508** [comment]
Still getting this issue. any solution?

**Comment by bhanuka96** [comment]
any solution

**Comment by msarkrish** [comment]
Any Update on this? I'm also facing this problem. I'm using Native code encrypting and decrypting the file.  In particular screen I decrypt all the encrypted files and displayed to the User. If i'm showing 1 file then it works fine. If i have more than 1 file then app will crash as Reply already submitted. I want to call the native method continuously in that screen. How to solve this?

**Comment by AmandaOlearczuk** [comment]
> Any Update on this? I'm also facing this problem. I'm using Native code encrypting and decrypting the file. In particular screen I decrypt all the encrypted files and displayed to the User. If i'm showing 1 file then it works fine. If i have more than 1 file then app will crash as Reply already submitted. I want to call the native method continuously in that screen. How to solve this?

I also would like to call native method (from a plugin) continuously and Im getting Reply already submitted..

**Comment by hamdikahloun** [comment]
I'm not able to reproduce the issue.

Hi @msarkrish @AmandaOlearczuk 

Can you please provide your `flutter doctor -v`, your `flutter run --verbos`e and a minimal complete reproducible code sample?

<details>
  <summary>Native code</summary>

```
public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success(android.os.Build.VERSION.RELEASE);
    } else if (call.method.equals("getCurrentTime")) {
      result.success("00:00");
    } else {
      result.notImplemented();
    }
  }
```
</details>

<details>
  <summary>Dart code</summary>

```
static const MethodChannel _channel =
      const MethodChannel('mychannel');

  static Future<Null> get platformVersion async {
    await _channel.invokeMethod('getPlatformVersion');
  }

  static Future<Null> get currentTime async {
    await _channel.invokeMethod('getCurrentTime');
  }

  static void get callTwoMethods {
    currentTime;
    platformVersion;
  }
```
</details>
<details>
  <summary>flutter doctor -v</summary>

```
[√] Flutter (Channel stable, 1.22.2, on Microsoft Windows [version 10.0.19041.572], locale fr-FR)
    • Flutter version 1.22.2 at D:\fork\flutter
    • Framework revision 84f3d28555 (2 weeks ago), 2020-10-15 16:26:19 -0700
    • Engine revision b8752bbfff
    • Dart version 2.10.2


[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at C:\Users\hamdi\AppData\Local\Android\Sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = C:\Users\hamdi\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[!] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code (version 1.50.1)
    • VS Code at C:\Users\hamdi\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.15.1

[√] Connected device (1 available)
    • STK L21 (mobile) • XMJ4C19B09007407 • android-arm64 • Android 10 (API 29)

! Doctor found issues in 1 category.
```

</details>

Thank you

**Comment by jcleng** [comment]
occasionally things go wrong

```shell
D/AndroidRuntime( 8753): Shutting down VM
E/AndroidRuntime( 8753): FATAL EXCEPTION: main
E/AndroidRuntime( 8753): Process: com.yunsheng.xuewl.app, PID: 8753
E/AndroidRuntime( 8753): java.lang.IllegalStateException: Reply already submitted
E/AndroidRuntime( 8753):        at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:139)
E/AndroidRuntime( 8753):        at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:238)
E/AndroidRuntime( 8753):        at com.wxwx.flutter_alibc.FlutterAlibcHandle$5.onFailure(FlutterAlibcHandle.java:209)
E/AndroidRuntime( 8753):        at com.alibaba.baichuan.android.trade.b.c$1.run(Unknown Source:10)E/AndroidRuntime( 8753):        at android.os.Handler.handleCallback(Handler.java:883)
E/AndroidRuntime( 8753):        at android.os.Handler.dispatchMessage(Handler.java:100)
E/AndroidRuntime( 8753):        at android.os.Looper.loop(Looper.java:224)
E/AndroidRuntime( 8753):        at android.app.ActivityThread.main(ActivityThread.java:7584)
E/AndroidRuntime( 8753):        at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 8753):        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:539)
E/AndroidRuntime( 8753):        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
D/JIGUANG-JCommon-P( 8753): [JCommonInterface] onEvent action:deviceinfo
W/jg_fr_pool_thre( 8753): type=1400 audit(0.0:73837): avc: denied { read } for name="u:object_r:deviceid_prop:s0" dev="tmpfs" ino=26685 scontext=u:r:untrusted_app_27:s0:c17,c257,c512,c768 tcontext=u:object_r:deviceid_prop:s0 tclass=file permissive=0
W/jg_fr_pool_thre( 8753): type=1400 audit(0.0:73838): avc: denied { read } for name="u:object_r:deviceid_prop:s0" dev="tmpfs" ino=26685 scontext=u:r:untrusted_app_27:s0:c17,c257,c512,c768 tcontext=u:object_r:deviceid_prop:s0 tclass=file permissive=0
E/libc    ( 8753): Access denied finding property "persist.radio.meid"
E/libc    ( 8753): Access denied finding property "ro.ril.oem.meid"
E/libc    ( 8753): Access denied finding property "ro.ril.miui.imei0"
E/libc    ( 8753): Access denied finding property "ro.ril.miui.imei"
E/libc    ( 8753): Access denied finding property "ro.ril.miui.imei1"
E/libc    ( 8753): Access denied finding property "ro.ril.oem.imei1"
W/jg_fr_pool_thre( 8753): type=1400 audit(0.0:73839): avc: denied { read } for name="u:object_r:core_deviceid_prop:s0" dev="tmpfs" ino=26641 scontext=u:r:untrusted_app_27:s0:c17,c257,c512,c768 tcontext=u:object_r:core_deviceid_prop:s0 tclass=file permissive=0
I/chatty  ( 8753): uid=10273(com.yunsheng.xuewl.app) identical 1 line
W/jg_fr_pool_thre( 8753): type=1400 audit(0.0:73841): avc: denied { read } for name="u:object_r:core_deviceid_prop:s0" dev="tmpfs" ino=26641 scontext=u:r:untrusted_app_27:s0:c17,c257,c512,c768 tcontext=u:object_r:core_deviceid_prop:s0 tclass=file permissive=0
E/libc    ( 8753): Access denied finding property "ro.ril.oem.imei2"
E/libc    ( 8753): Access denied finding property "persist.radio.imei1"
E/libc    ( 8753): Access denied finding property "persist.radio.imei2"
E/libc    ( 8753): Access denied finding property "vendor.cellular.ril.imsi"
W/sheng.xuewl.ap( 8753): Accessing hidden method Lcom/android/id/impl/IdProviderImpl;-><init>()V (greylist, reflection, allowed)
W/sheng.xuewl.ap( 8753): Accessing hidden method Lcom/android/id/impl/IdProviderImpl;->getOAID(Landroid/content/Context;)Ljava/lang/String; (greylist, reflection, allowed)
W/sheng.xuewl.ap( 8753): Accessing hidden method Lcom/android/id/impl/IdProviderImpl;->getVAID(Landroid/content/Context;)Ljava/lang/String; (greylist, reflection, allowed)
```

```shell
❯ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 1.22.0, on Microsoft Windows [Version 10.0.18363.1139], locale zh-CN)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.1)
[√] Android Studio (version 3.6)
[√] VS Code (version 1.50.1)
[√] Connected device (1 available)
```

**Comment by volgin** [comment]
Same error with a different plugin (AudioFilePlayer):

```
E/AndroidRuntime(18077): java.lang.IllegalStateException: Reply already submitted
E/AndroidRuntime(18077): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:139)
E/AndroidRuntime(18077): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.error(MethodChannel.java:243)
E/AndroidRuntime(18077): 	at com.google.flutter.plugins.audiofileplayer.AudiofileplayerPlugin.lambda$onLoad$1$AudiofileplayerPlugin(AudiofileplayerPlugin.java:377)
E/AndroidRuntime(18077): 	at com.google.flutter.plugins.audiofileplayer.-$$Lambda$AudiofileplayerPlugin$A6LKsuboaeEM7gfQPisQG3fhlSg.onRemoteLoadComplete(Unknown Source:11)
E/AndroidRuntime(18077): 	at com.google.flutter.plugins.audiofileplayer.RemoteManagedMediaPlayer.onError(RemoteManagedMediaPlayer.java:116)
E/AndroidRuntime(18077): 	at android.media.MediaPlayer$EventHandler.handleMessage(MediaPlayer.java:3440)
E/AndroidRuntime(18077): 	at android.os.Handler.dispatchMessage(Handler.java:107)
E/AndroidRuntime(18077): 	at android.os.Looper.loop(Looper.java:214)
E/AndroidRuntime(18077): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
E/AndroidRuntime(18077): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(18077): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E/AndroidRuntime(18077): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)

```

**Comment by hamdikahloun** [comment]
Hi @volgin  @bhanuka96 

From what I can see, the issue is related to a [audiofileplayer](https://github.com/google/flutter.plugins) and [FlutterLocation](https://github.com/Lyokone/flutterlocation) plugin rather than to Flutter itself .

```
E/AndroidRuntime(18077): 	at com.google.flutter.plugins.audiofileplayer.AudiofileplayerPlugin.lambda$onLoad$1$AudiofileplayerPlugin(AudiofileplayerPlugin.java:377)
E/AndroidRuntime(18077): 	at com.google.flutter.plugins.audiofileplayer.-$$Lambda$AudiofileplayerPlugin$A6LKsuboaeEM7gfQPisQG3fhlSg.onRemoteLoadComplete(Unknown Source:11)
E/AndroidRuntime(18077): 	at com.google.flutter.plugins.audiofileplayer.RemoteManagedMediaPlayer.onError(RemoteManagedMediaPlayer.java:116)
```

`E/AndroidRuntime(30172): 	at com.lyokone.location.FlutterLocation.onActivityResult(FlutterLocation.java:173)`

Thank you

**Comment by volgin** [comment]
@hamdikahloun 

Of course :), and I was able to resolve it. The point is that this exception should not crash the entire app. If someone runs into this problem in _any_ plugin, they may not be able to figure out what happens unless they catch it in their IDE - app crashing means that it cannot report this exception back to its developers.

**Comment by nateshmbhat** [comment]
any solution for this problem ? i actually have a use case which needs to make simultaneous method channel calls and fetch different results for corresponding calls.

**Comment by xster** [comment]
Seems like we were never able to reproduce it https://github.com/flutter/flutter/issues/29092#issuecomment-627627863. 

If anyone has a reduced reproduction project, we could take a look.

**Comment by adamshamsudeen** [comment]
@bhanuka96 Were you able to solve the issue?

**Comment by khiromoshito** [comment]
Hi.

I've arrived with this issue already, and fortunately I think I have found a solution for this. I actually assumed that maybe there are duplication in the success method of the MethodChannel.Result, and it could be the reason for the "already submitted" issue.

What I did was make a class that handles my requests instead of directly calling the submit method. Inside this class is a boolean flag that checks if the channel already submitted before. The code looks like this:

`class FlutterRequest {
    boolean hasSubmitted = false;
    MethodCall call;
    MethodChannel.Result submitter;

    FlutterRequest(MethodCall call, MethodChannel.Result submitter) {
        this.call = call;
        this.submitter = submitter;
    }

    public Object getArguments() {
        return call.arguments();
    }

    public String getMethod() {
        return call.method;
    }

    public void submit(Object message) {
        if(!this.hasSubmitted) {
            this.hasSubmitted = true;

            this.submitter.success(message);
        }
    }

    public void reportError(Object message) {
        if(!this.hasSubmitted) {
            this.hasSubmitted = true;

            this.submitter.error(message);
        }
    }

}`

So, I wrapped the request with this class and did all my methods and calls with it.

`final FlutterRequest myRequest = new FlutterRequest(call, result);`

If ever I want to call the original success method, I would instead call `myRequest.submit("message")` to make sure there will be no 'agains.' 

I haven't benchmarked this yet, but when I tried this code out, the errors did not show up anymore.
Hope it helps :)

**Comment by khiromoshito** [comment]
Hi.

I've arrived with this issue already, and fortunately I think I have found a solution for this. I actually assumed that maybe there are duplication in the success method of the MethodChannel.Result, and it could be the reason for the "already submitted" issue.

What I did was make a class that handles my requests instead of directly calling the submit method. Inside this class is a boolean flag that checks if the channel already submitted before. The code looks like this:

```

class FlutterRequest {
    boolean hasSubmitted = false;
    MethodCall call;
    MethodChannel.Result submitter;

    FlutterRequest(MethodCall call, MethodChannel.Result submitter) {
        this.call = call;
        this.submitter = submitter;
    }

    public Object getArguments() {
        return call.arguments();
    }

    public String getMethod() {
        return call.method;
    }

    public void submit(Object message) {
        if(!this.hasSubmitted) {
            this.hasSubmitted = true;

            this.submitter.success(message);
        }
    }

    public void reportError(Object message) {
        if(!this.hasSubmitted) {
            this.hasSubmitted = true;

            this.submitter.error(message);
        }
    }

}

```

So, I wrapped the request with this class and did all my methods and calls with it.

`final FlutterRequest myRequest = new FlutterRequest(call, result);`

If ever I want to call the original success method, I would instead call `myRequest.submit("message")` to make sure there will be no 'agains.' 

I haven't benchmarked this yet, but when I tried this code out, the errors did not show up anymore.
Hope it helps :)

**Comment by sidy-refyne** [comment]
It can be a problem related to hot reload as well. If you are going to define the result object as a field value on android side, It can happen.

**Comment by cedvdb** [comment]
> Seems like we were never able to reproduce it [#29092 (comment)](https://github.com/flutter/flutter/issues/29092#issuecomment-627627863).
> 
> If anyone has a reduced reproduction project, we could take a look.

It happens all the time with [sms autofill](https://github.com/jaumard/sms_autofill/issues/33) package when hot reloading.

the line that makes the error happen is this one:

```
  Future<String> get hint async {
    final String version = await _channel.invokeMethod('requestPhoneHint');
    return version;
  }
```

Basically when hot reloading multiple time, the hint to get user phone number appears multiple times and makes the app crash

**Comment by stuartmorgan** [comment]
I'm retitling to make it clearer what the original report was, and I'm going to hide a lot of off-topic discussion to make it less confusing.

To be very clear: It is not the case that any instance of `Reply already submitted` is this issue.
- If you see that error from a plugin **you are using**, you should report it **to the plugin author**, as a likely explanation for getting that error message is a bug in the plugin logic.
- If you are **a plugin author** and are seeing (or getting reports of) this message, you should audit your code to make sure you are never calling a `Result` response method more than once.
  - If you find that your logic is correct, and have multiple in-flight method channel requests at the same time, that would be where this issue would be relevant.

**Comment by ulmas97** [comment]
Just removing the 2nd result.success() solved the problem

**Comment by ollydixon** [comment]
Both packages ```image_picker``` and ```camera``` are now broken on Android because of this issue. Pretty darn critical bug.

**Comment by beylmk** [comment]
>   * If you find that your logic is correct, and have multiple in-flight method channel requests at the same time, that would be where this issue would be relevant.

@stuartmorgan Can you please explain what you mean by this part?

**Comment by stuartmorgan** [comment]
See the original issue description at the top of this page; the reported issue is that having two asynchronous method channel requests in flight *at the same time* could cause this error. (We haven't verified this yet, but that's the reported issue that's being tracked here.)

**Comment by blasten** [comment]
@gaaclarke was it intentional to register the handler for the [method channel twice] in (https://github.com/gaaclarke/double_result/blob/master/android/src/main/java/com/example/double_result/DoubleResultPlugin.java)?


cc @GaryQian will be taking a look

**Comment by gaaclarke** [comment]
> @gaaclarke was it intentional to register the handler for the [method channel twice] in (https://github.com/gaaclarke/double_result/blob/master/android/src/main/java/com/example/double_result/DoubleResultPlugin.java)?

@blasten I'm not sure.  I wrote this a long time ago.  Since it was me trying to reproduce a reported bug, it may be possible i was intentionally doing something wrong.

**Comment by Tiamenti** [comment]
I'm not sure if the problem is related to Flutter itself, but I ran into it too. I am using the laravel_echo 1.0.0-beta.1 package and I get a similar error when I try to call the `trigger` method.

**Flutter** ([Full code](https://pastebin.com/qFxDHufy))
`pusherClient.subscribe("orders").trigger("HelloEvent", "Hello");`

Logs
```
D/PusherClientPlugin(20885): [TRIGGER] client-HelloEvent
E/PusherClientPlugin(20885): Reply already submitted
W/System.err(20885): java.lang.IllegalStateException: Reply already submitted
W/System.err(20885): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:164)
W/System.err(20885): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:238)
W/System.err(20885): 	at com.github.chinloyal.pusher_client.pusher.PusherService.trigger(PusherService.kt:288)
W/System.err(20885): 	at com.github.chinloyal.pusher_client.pusher.PusherService.access$trigger(PusherService.kt:30)
W/System.err(20885): 	at com.github.chinloyal.pusher_client.pusher.PusherService$register$1.onMethodCall(PusherService.kt:60)
W/System.err(20885): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
W/System.err(20885): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:84)
W/System.err(20885): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:865)
W/System.err(20885): 	at android.os.MessageQueue.nativePollOnce(Native Method)
W/System.err(20885): 	at android.os.MessageQueue.next(MessageQueue.java:336)
W/System.err(20885): 	at android.os.Looper.loop(Looper.java:174)
W/System.err(20885): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
W/System.err(20885): 	at java.lang.reflect.Method.invoke(Native Method)
W/System.err(20885): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
W/System.err(20885): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): Failed to handle method call
E/MethodChannel#com.github.chinloyal/pusher_client(20885): java.lang.IllegalStateException: Reply already submitted
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:164)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.error(MethodChannel.java:243)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at com.github.chinloyal.pusher_client.pusher.PusherService.trigger(PusherService.kt:292)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at com.github.chinloyal.pusher_client.pusher.PusherService.access$trigger(PusherService.kt:30)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at com.github.chinloyal.pusher_client.pusher.PusherService$register$1.onMethodCall(PusherService.kt:60)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:84)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:865)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at android.os.MessageQueue.next(MessageQueue.java:336)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at android.os.Looper.loop(Looper.java:174)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E/MethodChannel#com.github.chinloyal/pusher_client(20885): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
E/DartMessenger(20885): Uncaught exception in binary message listener
E/DartMessenger(20885): java.lang.IllegalStateException: Reply already submitted
E/DartMessenger(20885): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:164)
E/DartMessenger(20885): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:253)
E/DartMessenger(20885): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:84)
E/DartMessenger(20885): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:865)
E/DartMessenger(20885): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/DartMessenger(20885): 	at android.os.MessageQueue.next(MessageQueue.java:336)
E/DartMessenger(20885): 	at android.os.Looper.loop(Looper.java:174)
E/DartMessenger(20885): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
E/DartMessenger(20885): 	at java.lang.reflect.Method.invoke(Native Method)
E/DartMessenger(20885): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E/DartMessenger(20885): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
E/flutter (20885): [ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: PlatformException(TRIGGER_ERROR, Trigger can only be called on private and presence channels., null, null)
E/flutter (20885): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:607:7)
E/flutter (20885): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:156:18)
E/flutter (20885): <asynchronous suspension>
E/flutter (20885): #2      Channel.trigger (package:pusher_client/src/pusher/channel.dart:68:5)
E/flutter (20885): <asynchronous suspension>
E/flutter (20885): 
```
